CMakeLists.txt

```diff
@@ -156,6 +156,7 @@ install(DIRECTORY contrib/HIP/include/ DESTINATION include/hipSYCL/contrib)
 
 install(PROGRAMS bin/syclcc DESTINATION bin)
 install(PROGRAMS bin/syclcc-clang DESTINATION bin)
+install(PROGRAMS bin/cmake-syclcc-launcher DESTINATION bin)
```

Comment:
Can you please elaborate why we need this launcher at all? Why can we not directly call syclcc?

Reply:
The launcher works around CMake's inflexibility when it comes to customizing compiler invocation. The compiler executable can only be set globally (and only from the command line) via the `CMAKE_CXX_COMPILER` variable, but not on a per-target or per-file basis. Changing it globally would mean that we can't control which targets are using syclcc and which targets are using the default system compiler. Currently this is solved by introducing the `SYCL` language in CMake which has syclcc set via `CMAKE_SYCL_COMPILER`, but has the drawbacks mentioned above.

I'll try explaining the launcher logic with an example. Say for some target, CMake generates the following compiler invocation:
```
/usr/bin/c++ -c -o sycl_code.o sycl_code.cc
```
Instead, since this is a SYCL source file, we want to invoke syclcc like so:
```
/usr/local/bin/syclcc --hipsycl-platform=cuda --hipsycl-gpu-arch=sm_75 -c -o sycl_code.o sycl_code.cc
```
Now we can't just edit that command line from within CMake itself, but we can define a compiler launcher via `set_property(RULE_LAUNCH_COMPILE)`. This mechanism is meant for tools like ccache, and it works by prepending the command line with a string provided in the CMake lists like so:
```
/path/to/launcher --launcher-args /usr/bin/c++ -c -o sycl_code.o sycl_code.cc
```
Now we can't just set `/usr/local/bin/syclcc` as the launcher since it does not know what to do with the `/usr/bin/c++` argument, and also we want to pass `--hipsycl` arguments that are not present in the original invocation for clang compatibility.

Instead, we introduce `cmake-syclcc-launcher`, which performs this command line transformation for us. After setting it as the compiler launcher, CMake invokes the following command:
```
/usr/local/bin/cmake-syclcc-launcher \
    --launcher-syclcc=/usr/local/bin/syclcc --launcher-cxx-compiler=/usr/bin/c++ \
    --hipsycl-platform=cuda --hipsycl-gpu-arch=sm_75 \
    /usr/bin/c++ -c -o sycl_code.o sycl_code.cc
```
The arguments tell the launcher to replace `/usr/bin/c++` with `/usr/local/bin/syclcc` and to additionally pass both `--hipsycl` parameters to syclcc, yielding the desired command from above. This substitution is only performed when `--launcher-cxx-compiler` matches the actual compiler executable specified in the command line from CMake, since the launcher will also be called for e.g. static library linking, which would invoke `ar` instead of `c++`.

Reply:
I see, that makes sense. Thanks for the in-depth explanation! 

I wonder if the launcher should reside in the `cmake` directory then, as it is an implementation detail for the cmake integration. Otherwise users installing hipSYCL globally would have it in `/usr/bin` which might not be desired.

Reply:
I agree. I canged the install script to move the launcher to `lib/cmake/hipSYCL/cmake-syclcc-launcher`. I have also moved the package config file `hipsycl-config.cmake` from `lib/cmake` to `lib/cmake/hipSYCL`. Having this additional subdirectory seems to be the "industry standard" anyway.

Note that this change may require users to change their `CMAKE_PREFIX_PATH`, if it has previously been set to e.g. `~/installs/hipSYCL/lib`, to `~/installs/hipSYCL/lib/cmake`. For a `/usr/local` install, this should not be needed though.

Reply:
That might break some projects that have cmake hipSYCL support such as SYCL-Academy or SYCL-Bench. But if this is the new industry standard we should go with it.

Reply:
SYCL-Bench compiles fine, but SYCL-Academy fails with #310, the reason being that the `CXX_STANDARD` setting was previously not forwarded to `SYCL` language targets.